Feat : OW-22 GlobalException 구축

Dongjin113 opened the pull request:
ApiResponse : 모든 api 응답 모델에서 사용가능하게 Generic Type으로 정의
 - 정상 api 응답 : 
         반환타입 : ApiResponse 
         return Type : new ApiResponse<>(ErrorCode, 응답보내줄 responseDto);
 - Error 발생 : 
         반환타입 : ApiResponse 
         return Type : ErrorResponse(ErrorCode, ErrorData)

ErrorData : 
 - Error 대한 상세내용이 더 필요하다면 객체를 생성하고 ErrorData 상속받아 사용
 - ErrorResponse의 생성자 파라미터에 Generic Type을 사용하지 않고, ErrorData 인터페이스로 타입을 제한하는게 안정성이 높다고 판단했다.

ErrorCode : 
 - 어떠한 에러들이 있는지 Enum Type으로 정의 message를 변경하고자 한다면 ErrorCode.반환에러타입.changeMessage(바꿀내용)으로 메시지의 내용만 변경 가능

CustomException : 
 - 도메인에서 Exception에 대한 정의가 필요하다면 CustomException 상속받아 정의

Dongjin113 committed:
Feat : OW-22 GlobalException 구축

ApiResponse : 모든 api 응답 모델에서 사용가능하게 Generic Type으로 정의
 - 정상 api 응답 :
     반환타입 : ApiResponse
     return Type : new ApiResponse<>(ErrorCode, 응답보내줄 response);
 - Error 발생 :
     반환타입 : ApiResponse
     return Type : ErrorResponse(ErrorCode, ErrorData)

ErrorData : Error 대한 상세내용이 더 필요하다면 객체를 생성하고 ErrorData 상속받아 사용

ErrorCode : 어떠한 에러들이 있는지 Enum Type으로 정의 message를 변경하고자 한다면 ErrorCode.반환에러타입.changeMessage(바꿀내용)으로 메시지의 내용만 변경 가능

CustomException : 도메인에서 Exception에 대한 정의가 필요하다면 CustomException 상속받아 정의

bstaran commented on back/src/main/java/com/ogjg/back/common/exception/ErrorCode.java:
```diff
@@ -0,0 +1,28 @@
+public enum ErrorCode {
+
+    SUCCESS(HttpStatus.OK,"200","OK"),
+    NOT_FOUND_USER(HttpStatus.NOT_FOUND, "코드는내맘대로에러","회원를 찾을 수 없습니다");
```
예시지만, 실제로 깃허브에 올라가는 코드이기 때문에 임시로라도 임시 에러 코드를 예시로 넣어보는 것은 어떠신가요?
- `"코드는내맘대로에러"`에서 `"4001"`(임시) 처럼 바꾸면 다른 사람이 보기에 어색하지 않을 것 같습니다!

그리고 `,`뒤에 띄어쓰기와 `"를"` 오타 수정 부탁드립니다!

```suggestion
    NOT_FOUND_USER(HttpStatus.NOT_FOUND, "4001", "회원을 찾을 수 없습니다");
```

Dongjin113 committed:
Feat : OW-22 GlobalException Validation Exception Handler 추가

- @Valid에서 발생하는 데이터 검증 오류 Handler 추가
- 이전 코드들의 띄어쓰기 등 가독성 개선
- ErrorCode code 정의

dunowljj commented on back/src/main/java/com/ogjg/back/common/response/ErrorResponse.java:
```diff
@@ -0,0 +1,15 @@
+import com.ogjg.back.common.exception.ErrorCode;
+import com.ogjg.back.common.exception.ErrorData;
+
+public class ErrorResponse extends ApiResponse<Object> {
```
Object가 아닌 ErrorData로 제네릭 타입을 지정해주시면 생성자 파라미터를 ErrorData로 사용하도록 강제할 수 있을거 같습니다. 어떻게 생각하시나요?

Dongjin113 committed:
Feat : OW-22 GlobalException ErrorResponse
- ApiResponse 들어오는 T type ErrorResponse에서 ErrorData만 넘길수 있도록 강제하기

bstaran reviewed (approved):
오 너무 좋습니다. 고생하셨습니다!

Dongjin113 committed:
Feat : OW-22 GlobalException ExceptionControllerAdvice

- customValidationHandler ValidExceptionHandler 메서드명 변경 customValidationHandler -> validationHandler

dunowljj reviewed (approved):
고생하셨습니다~
